File: C++/Searching/InterpolationSearch.cpp
#include <iostream>

// Interpolation Search algorithm
int interpolationSearch(int arr[], int n, int x) {
    int low = 0, high = (n - 1);
 
    while (low <= high && x >= arr[low] && x <= arr[high]) {
        // Estimate the position of x in the array
        int pos = low + (((double)(high - low) / (arr[high] - arr[low])) * (x - arr[low]));
 
        // If x is found at position pos
        if (arr[pos] == x)
            return pos;
 
        // If x is greater, ignore the left half
        if (arr[pos] < x)
            low = pos + 1;
        // If x is smaller, ignore the right half
        else
            high = pos - 1;
    }
 
    // Element not found
    return -1;
}

int main() {
    int arr[] = {2, 4, 6, 8, 10, 12, 14, 16, 18, 20};
    int n = sizeof(arr) / sizeof(arr[0]);
    int x = 14;
    
    // Perform interpolation search
    int result = interpolationSearch(arr, n, x);
    
    // Check if the element was found or not
    if (result == -1)
        std::cout << "Element not found." << std::endl;
    else
        std::cout << "Element found at index " << result << "." << std::endl;
 
    return 0;
}

File: Python/Searching/interpolation_search.py

#----------------------------

#problem - Given a sorted integer array and a target, 
# determine if the target exists in the array or not using an interpolation search algorithm. 
# If the target exists in the array, return the index of it.

#------------------------

#it is same as binary search but it use this formula to calculate mid -
 # mid = low + ((target – A[low]) * (high – low) / (A[high] – A[low]));

#-----------------------

#variables - 
#start - starting index also know as low or low (low most element )
#end - ending index also know as high or right ( right most element)
#target - element to be search in the array
#size - size of the array 
  
#-----------------------------


def interpolationSearch(array, start, end, target):

	# Since array is sorted, an element present
    # search space is array[low…high]
	if (start <= end and target >= arr[start] and target <= array[end]):

		# Probing the position with keeping
		# uniform distribution in mind.
		mid = start + ((end - start) // (array[end] - array[start]) *
					(target - arr[start]))

		# #target value found
		if array[mid] == target:
			return mid

		# If x is larger, x is in right subarray
		if array[start] < target:
			return interpolationSearch(array, mid + 1,
									end, target)

		# If x is smaller, x is in left subarray
		if array[mid] > target:
			return interpolationSearch(array, start,
									mid - 1, target)
	return -1

# Driver code


# Array of items in which
# search will be conducted
arr = [10, 12, 13, 16, 18, 19, 20,
	21, 22, 23, 24, 33, 35, 42, 47]
n = len(arr)

# Element to be searched
x = 18
index = interpolationSearch(arr, 0, n - 1, x)

if index != -1:
	print("Element found at index", index)
else:
	print("Element not found")


# code is contributed by rajendra mandliya